mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java

```diff
@@ -0,0 +1,161 @@
+import org.apache.iceberg.BaseMetastoreTableOperations;
+import org.apache.iceberg.PartitionSpecParser;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.SchemaParser;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.hive.HiveTableOperations;
+import org.apache.iceberg.mr.Catalogs;
+import org.apache.iceberg.mr.InputFormatConfig;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class HiveIcebergMetaHook implements HiveMetaHook {
+  private static final Logger LOG = LoggerFactory.getLogger(HiveIcebergMetaHook.class);
+  private static final Set<String> PARAMETERS_TO_REMOVE = Stream
+      .of(InputFormatConfig.TABLE_SCHEMA, InputFormatConfig.PARTITION_SPEC, Catalogs.LOCATION, Catalogs.NAME)
+      .collect(Collectors.toCollection(HashSet::new));
+  private static final Set<String> PROPERTIES_TO_REMOVE = Stream
+      .of(InputFormatConfig.HIVE_DELETE_BACKING_TABLE, "storage_handler", "EXTERNAL")
+      .collect(Collectors.toCollection(HashSet::new));
+
+  private final Configuration conf;
+  private Table icebergTable = null;
+  private Properties catalogProperties;
+  private boolean deleteIcebergTable;
+
+  public HiveIcebergMetaHook(Configuration conf) {
+    this.conf = conf;
+  }
+
+  @Override
+  public void preCreateTable(org.apache.hadoop.hive.metastore.api.Table hmsTable) {
+    catalogProperties = getCatalogProperties(hmsTable);
+    try {
+      icebergTable = Catalogs.loadTable(conf, catalogProperties);
+
+      Preconditions.checkArgument(catalogProperties.getProperty(InputFormatConfig.TABLE_SCHEMA) == null,
+          "Iceberg table already created - can not use provided schema");
+      Preconditions.checkArgument(catalogProperties.getProperty(InputFormatConfig.PARTITION_SPEC) == null,
+          "Iceberg table already created - can not use provided partition specification");
+
+      LOG.info("Iceberg table already exists {}", icebergTable);
+    } catch (NoSuchTableException nte) {
+      String schemaString = catalogProperties.getProperty(InputFormatConfig.TABLE_SCHEMA);
+      Preconditions.checkNotNull(schemaString, "Please provide a table schema");
+      // Just check if it is parsable, and later use for partition specification parsing
+      Schema schema = SchemaParser.fromJson(schemaString);
+
+      String specString = catalogProperties.getProperty(InputFormatConfig.PARTITION_SPEC);
+      if (specString != null) {
+        // Just check if it is parsable
+        PartitionSpecParser.fromJson(schema, schemaString);
+      }
+
+      // Allow purging table data if the table is created now and not set otherwise
+      if (hmsTable.getParameters().get(InputFormatConfig.HIVE_DELETE_BACKING_TABLE) == null) {
+        hmsTable.getParameters().put(InputFormatConfig.HIVE_DELETE_BACKING_TABLE, "TRUE");
```

Comment:
Can you explain this option?

Reply:
I can see 2 different use-cases for Hive tables Iceberg tables:
1. Table is created from Hive. When we drop the table we want to drop the underlying Iceberg table as well
2. Table is created outside of Hive but want to read it from Hive as well. We create the Hive table above the Iceberg table but when we drop the table we do not want to drop the underlying Iceberg table.

The proposed solution is:
- Default behavior: When the table is created from Hive and it needs to create the underlying Iceberg table then Hive will master the data. If the underlying Iceberg table is already there then Hive will not master the data. When the Hive table is dropped then the underlying Iceberg table is only dropped if Hive is mastering the data.
- HIVE_DELETE_BACKING_TABLE could override the default behavior when dropping the table

There is 1 exception: If the HiveCatalog is used it is obviously not possible to keep the table if we drop the HMS table. That is why this check is needed.

Reply:
Would it work to drop the table if it is a HMS table and not if it uses a non-Hive catalog?

Reply:
Or alternatively, could we use the existing `EXTERNAL` option? I think that does basically the same thing for data in paths that Hive points to.

Reply:
The `EXTERNAL`, `MANAGED` definition shifted a little bit in Hive lately.
Previously the only real difference was that when the table was dropped, the data of `EXTERNAL` tables was not removed, while the data of the `MANAGED` tables were cleaned up. Since Hive 3 `MANAGED` means "do not touch my directories because there will be a data loss", and even for `EXTERNAL` tables we have a new property `external.table.purge`, which means that the data should be cleaned up if the table is dropped.

Reply:
I don't quite understand the new definition of `MANAGED` that you're talking about, but I think we should try to make it work. We could still respect `external.table.purge`. I'll try to find some docs on the new definition.

Reply:
Also: if we know that Hive is creating the table, we should definitely set the correct storage handler and serde. Tables that are created by Spark might be used by Hive (and should depend on hive-site.xml) but we can assume that tables created by Hive have the Iceberg library present and can be set up automatically, I think.

Reply:
Maybe this doc can help understand the new `MANAGED` tables: https://docs.cloudera.com/HDPDocuments/HDP3/HDP-3.1.5/using-hiveql/content/hive_hive_3_tables.html

Reply:
It seems like the distinction between external and managed still fits. @omalley, what do you think about this?

Reply:
Since managed means that it is hive-owned and in hive acid layout, it doesn't seem like we should have iceberg tables that hive thinks are managed.

It doesn't seem like we should have a separate property for deleting the external table's data and should just use the **external.table.purge** table property that Hive already uses. It probably makes sense to also delete the data when:

- We are using the HiveCatalog.
- The iceberg location property is set.

since once the Hive metadata is deleted, it won't be recoverable. Rather than having separate logic, it might make sense to have the HiveCatalog set that property when it creates a table.

Reply:
> managed means that it is hive-owned and in hive acid layout

Thanks! I didn't realize that managed now means that it uses the Hive ACID layout.

I agree that if we can reuse a property, then we should. I like the idea to use `external.table.purge`.

I would set that automatically when creating a table from Hive, but I don't think it makes sense to drop the table data automatically based on using `HiveCatalog`. Leaving metadata in place to be cleaned up by something like `RemoveOrphanFilesAction` is nice so that you can un-delete tables when a user dropped one accidentally. For example, we've seen lots of cases where tables in the wrong db are dropped because the user thought they had run `USE`.

I'd say let's keep it simple and just use the existing Hive property.

Reply:
Moved to using the `external.table.purge`, but realized that calling Catalogs.dropTable for HiveCatalog tables is not enough, since the table had been already dropped from the HMS, so the HiveCatalog will not find it.

Reply:
Could we detect this in the pre-delete method and set a flag so the post-delete method can purge?

Reply:
I was thinking around these lines.
The problem is that when we drop the table we drop the current snapshot as well.
We can access everything through Table object but for removing the data / metadata files, we need some more info. I know of `CatalogUtil.dropTableData(FileIO io, TableMetadata metadata)`, but for that we would need TableMetadata. I might be able to store the snapshot location and later use that, but it is quite late here and this needs more digging 😄

Reply:
[`StaticTableOperations`](https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/StaticTableOperations.java) might help here. We use that to inspect metadata that has already expired in the `ExpireSnapshotsAction`. That allows you to create a static table based on an older metadata location.

Reply:
Good pointers! Big thanks!
Found solution which looks nice 😄